### PR TITLE
Add support to function keys F1-F12, change TERM=xterm-256color

### PR DESCRIPTION
**Pull request by lxsang**
Add basic function keys F1-F12 support #60 .

### CODE REVIEW

**Comment by Rinzwind** [comment]
It would seem better to use [‘xterm’](https://invisible-island.net/ncurses/terminfo.src.html#tic-xterm) for ‘TERM’ rather than [‘xterm-256color’](https://invisible-island.net/ncurses/terminfo.src.html#tic-xterm-256color) as there’s no support yet for using 256 colors. When using ‘xterm-256color’, the function key row in GNU midnight commander doesn’t show the function key numbers: it shows  instead of . It does show the key numbers when ‘xterm’ is used.

A problem that occurs with both ‘xterm’ and ‘xterm-256color’ is that GNU ‘screen’ no longer shows its startup message (with the version number and license) like it shows with ‘TERM’ set to [‘linux’](https://invisible-island.net/ncurses/terminfo.src.html#tic-linux).

For `#processFunctionKey:` on TerminalEmulatorMorph, it might make sense to use a class variable with a dictionary to map the function keys to the byte arrays for the `downcallAll:` message. For the keys of the dictionary, KMSingleKeyCombination instances could be used, which can be gotten from a KeyboardKey through `#asKeyCombination`, which can also be sent to a KeyboardEvent.

**Comment by lxsang** [comment]
> It would seem better to use [‘xterm’](https://invisible-island.net/ncurses/terminfo.src.html#tic-xterm) for ‘TERM’

Yes, it is better use the term type string provided by the `TerminalEmulator...` class used, fixed in the last commit.

> A problem that occurs with both ‘xterm’ and ‘xterm-256color’ is that GNU ‘screen’ no longer shows its startup message (with the version number and license) like it shows with ‘TERM’ set to [‘linux’](https://invisible-island.net/ncurses/terminfo.src.html#tic-linux).

The Mac terminal don't have this problem with TERM set to xterm-256color, i don't know what causes this...

> For #processFunctionKey: on TerminalEmulatorMorph, it might make sense to use a class variable with a dictionary to map..

Agree on a dedicated class method that defines a function key map, however, the use of `KMSingleKeyCombination` as key is a bit overkill as that makes the code more complicate. I prefer the use of Symbol (e.g. `#F1`) which is easier to understand.

**Comment by Rinzwind** [comment]
I tried to debug the GNU ‘screen’ problem. It is related to the send of `#setWidth:` in `#decrmSingle:` on TerminalEmulatorVT102. When the send is replaced by `self`, the problem no longer occurs. It might be due to the send causing two sends of `#setWinsize:` with different arguments. Try the following: open a Terminal window, put a breakpoint at the start of `#setWinsize:` on PTerm, then execute `printf '\33[?3l'` in the Terminal window. Two debuggers are opened, in one the argument is `81@21` and in the other it is `79@21`.

**Comment by Rinzwind** [comment]
Note that the [’xterm’ manual](https://invisible-island.net/xterm/manpage/xterm.html) says the DECCOLM escape sequence is ignored unless the following option is given:

> **-132** Normally, the VT102 DECCOLM escape sequence that switches between 80 and 132 column mode is ignored. This option causes the DECCOLM escape sequence to be recognized, and the xterm window will resize appropriately.

**Comment by Rinzwind** [comment]
While the two sends of `#setWinsize:` with different arguments is something that should be fixed, it does not seem to be the cause of the GNU ‘screen’ problem. A similar problem occurs in xterm if the option `-132` is used: after using `printf '\33[8;25;100;t'` to resize the window to 25 lines and 100 columns, when ‘screen’ is started it shows a message line “Aborted because of window size change” instead of the startup message. There’s a reference to this problem in [Debian’s system-wide ‘screenrc’](https://salsa.debian.org/debian/screen/-/blob/a80f2111fdd5e29b5604a68b82ddf5b6c726618e/debian/screenrc#L80-83) file which has a [‘termcapinfo’ command](https://www.gnu.org/software/screen/manual/screen.html#Termcap-Syntax) that changes ‘is’ for ‘xterm’:

```
# Change the xterm initialization string from is2=\E[!p\E[?3;4l\E[4l\E>
# (This fixes the "Aborted because of window size change" konsole symptoms found
#  in bug #134198)
termcapinfo xterm 'is=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;4;6l'
```

The example file [‘etcscreen’ in the ‘screen’ repository](https://git.savannah.gnu.org/cgit/screen.git/tree/src/etc/etcscreenrc?id=c184c6ec27683ff1a860c45be5cf520d896fd2ef#n63) has similar commands to avoid the use of DECCOLM:

```
# we do not want the width to change to 80 characters on startup:
# on suns, /etc/termcap has :is=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;3;4;6l:
termcap xterm 'is=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;4;6l'
terminfo xterm 'is=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;4;6l'
```

There doesn’t seem to be a system-wide ‘screenrc’ on macOS (the file `/usr/local/etc/screenrc` referred to in the man page on macOS v13.6.7 doesn’t exist) hence ‘screen’ uses the standard value for ‘is’ given by the following:

```
$ printf 'is=%q\n' "$(tput is)"
is=$'\E[!p\E[?3;4l\E[4l\E>'
```

I think this can be dealt with separately from this pull request. The setting of the window size should be fixed so that the number of columns is correctly set to 80 instead of 81 or 79. The default value for `allow132` in `#initializeTeletype:` on TerminalEmulatorMorph should maybe be changed to false.

File: PTerm-UI/TerminalEmulatorMorph.class.st [hunk 1]
```diff
@@ -1255,44 +1273,61 @@ TerminalEmulatorMorph >> preferredExtent [
...
 	"Handle VT100 control key"
 	(evt controlKeyPressed)
 			ifTrue: [
-				(char = 0) ifTrue:[^self].
-				(char = 32) ifTrue: [ ^ down downcall: 0 ].
-				(char between: 65 and: 93) ifTrue: [ char := char + 32 ].
-				(char between: 97 and: 125) ifTrue: [ ^ down downcall: char - 96 ]
+				(char = 0) ifTrue:[^true].
+				(char = 32) ifTrue: [ down downcall: 0. ^true ].
+				(char between: 65 and: 93) ifTrue: [ down downcall: char + 32. ^true].
```
Comment (Rinzwind) [inline PTerm-UI/TerminalEmulatorMorph.class.st:1322]: The change to the case for `char` being between 65 and 93 doesn’t take into account that previously 96 would be subtracted after adding 32.

**Review by Rinzwind — CHANGES_REQUESTED** [review]
Something that does need to be dealt with in this pull request: there’s a small mistake in `#processModifierKey:`, see the comment below.

**Comment by lxsang** [comment]
Yes, my bad. Fixed now

**Review by Rinzwind — APPROVED** [review]
Approved.

One thing to note: TerminalType isn’t used in `#openPTerm:` for TerminalEmulator, it always uses TerminalEmulatorXterm.

I took a look at the [original Squeak goodie](http://squeakvm.org/unix/goodies.html#html/goodies.m4.2.0) using the [Squeak 3.8 image hosted at squeak.js.org](https://squeak.js.org/run/#url=https://freudenbergs.de/vanessa/squeakjs&files=[Squeak3.8.1-6747full.image,Squeak3.8.1-6747full.changes,SqueakV3.sources]&swapButtons=true). The ‘telnet.301.cs’ part can be loaded through the following snippet (for the question about creating a better fixed-width font, select ‘Yes’). The Teletype window menu allows configuring various preferences, including which terminal type to use (see the screenshot below).

```smalltalk
HTTPSocket httpFileIn: 'http://squeakvm.org/unix/goodies/telnet.301.cs'.
AbstractFont compile: 'isMonospaced ^ false'.
(Smalltalk at: #TeletypeWindow) open.
```

**Pull request merged**